python/rapidsmpf/rapidsmpf/config.pyx

```diff
@@ -0,0 +1,195 @@
+    return ret
+
+
+cdef class Options:
+    """Initialize an Options object with a dictionary of string options.
+
+    Parameters
+    ----------
+    options_as_strings
+        A dictionary representing option names and their corresponding values.
+    """
+    def __cinit__(self, options_as_strings):
+        cdef unordered_map[string, string] opts
+        for key, val in options_as_strings.items():
+            opts[str.encode(key)] = str.encode(val)
+        with nogil:
+            self._handle = cpp_Options(move(opts))
+
+    def get_or_assign(self, str key, parser_type, default_value):
+        """
+        Get the value of the given key, or assign and return a default value.
+
+        When a key is accessed for the first time, its option value (as a string)
+        is parsed using the `parser_type`. If the key has no option value,
+        `default_value` is assigned to the key and returned.
+
+        The type of the returned value is determined by the `parser_type` argument,
+        which must be one of the supported Python types: bool, int, float, or str.
+        The `default_value` is cast to the corresponding C++ type before being passed
+        to the underlying options system.
+
+        TODO: Implement Python bindings to `rapidsmpf::config::Options::get` that
+        support Python objects. This will make it possible to store and retrieve any
+        kind of Python objects, not just bool, int, float, and str. The downside is
+        that those Python objects will not be accessible to C++.
+
+        Note
+        ----
+        Once a key has been accessed with a particular `parser_type`, subsequent
+        calls to `get_or_assign` on the same key must use the same `parser_type`.
+        Using a different `parser_type` for the same key will result in a `ValueError`.
+        The first parsing determines the value type associated with that key.
+
+        Parameters
+        ----------
+        key
+            The key to look up or assign in the configuration.
+        parser_type
+            The expected type of the result. Must be one of: bool, int, float, str.
+        default_value
+            The value to assign and return if the key is not found. Its type must
+            match `parser_type`.
+
+        Returns
+        -------
+        The value associated with the key, parsed and cast to the given type,
+        or the assigned default value if the key did not exist.
+
+        Raises
+        ------
+        ValueError
+            If the type of `default_value` is not supported or does not match
+            `parser_type`.
+        """
+        if issubclass(parser_type, bool):
+            return _options_get_using_parse_string[bool_t](
+                self._handle, str.encode(key), bool(default_value)
+            )
+        elif issubclass(parser_type, int):
+            return _options_get_using_parse_string[int64_t](
+                self._handle, str.encode(key), int(default_value)
+            )
```

Comment:
Given Python `int` can have arbitrary precision, should we check for its limits and raise an error if it's beyond to avoid over/underflow?

Reply:
That is already done, added a test: eb63be63d3829546394eacef81c6120891176e57

Reply:
Thanks for adding the test. But where does that happen? I see the test is passing which proves you're right, I don't know/see how this happens.

Reply:
Cython raises OverflowError, added a test for that as well :)